NorthstarDLL/mods/modsavefiles.cpp

```diff
@@ -0,0 +1,515 @@
+				// tried to use m.try_lock(), but it's unreliable, it seems.
+			}
+			catch (std::exception ex)
+			{
+				spdlog::error("DELETE FAILED!");
+				m.get().unlock();
+				spdlog::error(ex.what());
+			}
+		});
+
+	deleteThread.detach();
+}
+
+bool ContainsInvalidChars(std::string str)
+{
+	// we don't allow null characters either, even if they're ASCII characters because idk if people can
+	// use it to circumvent the file extension suffix.
+	return std::any_of(str.begin(), str.end(), [](char c) { return c == '\0'; });
+}
+// I don't know why I have to pass the string path as a separate parameter
+// I don't want to know why I have to pass the string path as a separate parameter
+// but if I don't ASCII check fails. So here we are.
+bool IsPathSafe(const std::string param, fs::path dir)
+{
+	auto const normRoot = fs::weakly_canonical(dir);
+	auto const normChild = fs::weakly_canonical(dir / param);
+
+	auto itr = std::search(normChild.begin(), normChild.end(), normRoot.begin(), normRoot.end());
+	// we return if the file is safe (inside the directory) and uses only ASCII chars in the path.
+	return itr == normChild.begin() && std::none_of(
+										   param.begin(),
+										   param.end(),
+										   [](char c)
+										   {
+											   unsigned char unsignedC = static_cast<unsigned char>(c);
+											   return unsignedC > 127 || unsignedC < 0;
+										   });
+}
+
+// void NSSaveFile( string file, string data )
+ADD_SQFUNC("void", NSSaveFile, "string file, string data", "", ScriptContext::SERVER | ScriptContext::CLIENT | ScriptContext::UI)
+{
+	Mod* mod = g_pSquirrel<context>->getcallingmod(sqvm);
+	if (mod == nullptr)
+	{
+		g_pSquirrel<context>->raiseerror(sqvm, "Has to be called from a mod function!");
+		return SQRESULT_ERROR;
+	}
+
+	fs::path dir = savePath / fs::path(mod->m_ModDirectory).filename();
+	std::string fileName = g_pSquirrel<context>->getstring(sqvm, 1);
+	if (!IsPathSafe(fileName, dir))
+	{
+		g_pSquirrel<context>->raiseerror(
+			sqvm,
+			fmt::format(
+				"File name invalid ({})! Make sure it does not contain any non-ASCII character, and results in a path inside your mod's "
+				"save folder.",
+				fileName,
+				mod->Name)
+				.c_str());
+		return SQRESULT_ERROR;
+	}
+
+	std::string content = g_pSquirrel<context>->getstring(sqvm, 2);
+	if (ContainsInvalidChars(content))
+	{
+		g_pSquirrel<context>->raiseerror(
+			sqvm, fmt::format("File contents may not contain NUL/\\0 characters! Make sure your strings are valid!", mod->Name).c_str());
+		return SQRESULT_ERROR;
+	}
+
+	fs::create_directories((dir / fileName).parent_path());
+	// this actually allows mods to go over the limit, but not by much
+	// the limit is to prevent mods from taking gigabytes of space,
+	// this ain't a cloud service.
+	if (GetSizeOfFolderContentsMinusFile(dir, fileName) + content.length() > MAX_FOLDER_SIZE)
+	{
+		// tbh, you're either trying to fill the hard drive or use so much data, you SHOULD be congratulated.
```

Comment:
I'd swap sarcastic comments with what's actually happening / **why** an error is thrown.

The **what** can usually be inferred from code (although additional comments can help, keep in mind that a lot of first time contributors lack coding knowledge so a few additional comments don't hurt), so comments should be used to cover the **why** part.